File: 05 - SOLID Principles/Liskov Substitution Principle/Java Code/FollowsLSP.java
// =============================================================================
// LISKOV SUBSTITUTION PRINCIPLE — the FIX (Java twin of follows-lsp.cpp).
// Split by capability: only accounts that can really withdraw implement
// Withdrawable, and payBill() asks for exactly that. An FD won't compile there.
//
// Build & run:
//   javac FollowsLSP.java
//   java FollowsLSP
// =============================================================================
public class FollowsLSP {

    static class Account {
        protected double balance = 0;
        void deposit(double a) { this.balance += a; }
        double getBalance() { return this.balance; }
    }

    // A separate capability. Only accounts that can really do this implement it.
    interface Withdrawable {
        void withdraw(double amount);
    }

    static class SavingsAccount extends Account implements Withdrawable {
        public void withdraw(double amount) {
            if (amount > this.balance) throw new IllegalStateException("insufficient funds");
            this.balance -= amount;
        }
    }

    // An FD IS-A Account, but is simply NOT Withdrawable — the type says so.
    static class FixedDepositAccount extends Account {
        private int maturityMonths;
        FixedDepositAccount(int months) { this.maturityMonths = months; }
        int getMaturityMonths() { return this.maturityMonths; }
    }

    static void payBill(Withdrawable source, double amt) {
        source.withdraw(amt);
    }

    public static void main(String[] args) {
        SavingsAccount sav = new SavingsAccount();
        sav.deposit(5000);
        payBill(sav, 1200);                       // fine — Savings IS Withdrawable
        System.out.println("Savings balance after bill: Rs " + sav.getBalance());

        FixedDepositAccount fd = new FixedDepositAccount(12);
        fd.deposit(100000);
        System.out.println("FD balance: Rs " + fd.getBalance()
                + " (matures in " + fd.getMaturityMonths() + " months)");
        // payBill(fd, 100);  // <-- would NOT compile: FixedDepositAccount isn't Withdrawable
        System.out.println("Compiler refuses payBill(fd, ...): the LSP trap is gone.");
    }
}

File: 05 - SOLID Principles/Liskov Substitution Principle/C++ Code/follows-lsp.cpp
// =============================================================================
// LISKOV SUBSTITUTION PRINCIPLE — the FIX.
// Split the hierarchy by CAPABILITY. Only accounts that truly support
// withdrawal implement a Withdrawable interface, and payBill() asks for exactly
// that. Handing it a FixedDeposit becomes a COMPILE-TIME error, not a crash.
//
// Build & run:
//   g++ -std=c++14 follows-lsp.cpp -o follows
//   ./follows
// =============================================================================
#include <iostream>
#include <stdexcept>
#include <string>

using namespace std;

// Base = what EVERY account can honestly do.
class Account {
protected:
    double balance;
public:
    Account() {
        this->balance = 0;
    }
    void deposit(double a) { this->balance += a; }
    double getBalance() { return this->balance; }
    virtual ~Account() {}
};

// A separate capability. Only accounts that can really do this implement it.
class Withdrawable {
public:
    virtual void withdraw(double amount) = 0;
    virtual ~Withdrawable() {}
};

class SavingsAccount : public Account, public Withdrawable {
public:
    void withdraw(double amount) {
        if (amount > this->balance) throw runtime_error("insufficient funds");
        this->balance -= amount;
    }
};

// An FD IS-A Account, but it is simply NOT Withdrawable — and the type says so.
class FixedDepositAccount : public Account {
private:
    int maturityMonths;
public:
    FixedDepositAccount(int months) {
        this->maturityMonths = months;
    }
    int getMaturityMonths() { return this->maturityMonths; }
};

// This function asks for exactly the capability it needs. It can NEVER be handed
// something that can't withdraw — the mistake is caught at compile time.
void payBill(Withdrawable* source, double amt) {
    source->withdraw(amt);
}

int main() {
    SavingsAccount* sav = new SavingsAccount();
    sav->deposit(5000);
    payBill(sav, 1200);                       // fine — Savings IS Withdrawable
    cout << "Savings balance after bill: Rs " << sav->getBalance() << endl;

    FixedDepositAccount* fd = new FixedDepositAccount(12);
    fd->deposit(100000);
    cout << "FD balance: Rs " << fd->getBalance()
         << " (matures in " << fd->getMaturityMonths() << " months)" << endl;
    // payBill(fd, 100);  // <-- would NOT compile: FixedDepositAccount isn't Withdrawable
    cout << "Compiler refuses payBill(fd, ...): the LSP trap is gone." << endl;

    delete sav;
    delete fd;
    return 0;
}
